File: client/src/app/_models/Product.ts
import { Review } from "./Review";

export interface ProductResponse {
  items: Product[],
  page: number,
  pageSize: number,
  totalCount: number
}

export interface ProductRequest {
  id: number;
  name: string;
  originalPrice: number;
  salePrice: number;
  quantity: number;
  description: string;
  subCategoryId: number;
  categoryId: number;
  brandId: number;
  isCreate: boolean;
  colors: ColorRequest[];
  productSpecifications: ProductSpecification[];
}

export interface PhotoUpdateRequest {
  itemId: number;
  photoCollection: FileList;
  idsToRemove: number[];
}

export interface SetMainPhotoRequest {
  itemId: number;
  photoId: number;
}

export interface Product {
  [key: string]: any;
  
  id: number;
  name: string;
  description: string;
  productRating: number;
  totalReviews: number;
  brand: Brand;
  originalPrice: number;
  subCategoryId: number;
  categoryId: number;
  salePrice: number;
  quantity: number;
  colors: Color[];
  productImages: ProductImage[];
  mainImageUrl: string;
  productSpecifications: ProductSpecification[];
  reviews: Review[];
}

export interface Brand {
  id: number;
  brandName: string;
  brandDescription: string;
}

export interface ProductImage {
  id: number;
  url: string;
  isMain: boolean;
}

interface ProductSpecification {
  name: string;
  value: string;
}

interface ColorRequest {
  value: string;
}

export interface Color {
  id: number;
  value: string;
}

File: API/Models/Product/Product.cs
using Microsoft.EntityFrameworkCore;
using System.ComponentModel.DataAnnotations;

namespace API.Models.ProductModel
{
    public class Product
    {
        [Key]
        public int Id { get; set; }

        [MaxLength(150)]
        public string Name { get; set; }

        [Precision(16, 2)]
        public decimal OriginalPrice { get; set; }

        [Precision(16, 2)]
        public decimal SalePrice { get; set; }

        public int Quantity { get; set; }

        public List<ProductColor> Colors { get; set; }

        public string Description { get; set; }

        [Range(10, 50)]
        public double ProductRating { get; set; } = 0;

        public int TotalReviews { get; set; } = 0;
        public DateTime CreatedAt { get; set; }
        public bool IsPlaceholder { get; set; } = false;

        public int SubCategoryId { get; set; }
        public ProductSubCategory SubCategory { get; set; }

        [MaxLength(250)]
        public string MainImageUrl { get; set; }

        public List<ProductImage> ProductImages { get; set; } = new();
        public List<ProductSpecification> ProductSpecifications { get; set; } = new();
        public List<Review> Reviews { get; set; } = new();

        public int BrandId { get; set; }
        public Brand Brand { get; set; }

    }
}
